### ISSUE TITLE
Sentinel-2 Driver

### ISSUE THREAD

**Issue by ircwaves** [Applied-GeoSolutions/gips#200]
Implement a driver for the Sentinel-2 satellites. 

Functionality should include:
1. archiving of S2A (and eventually S2B) gridded assets
    + [asset naming convention](https://earth.esa.int/web/sentinel/user-guides/sentinel-2-msi/naming-convention)
    + [tile grid](https://sentinel.esa.int/documents/247904/1955685/S2A_OPER_GIP_TILPAR_MPC__20151209T095117_V20150622T000000_21000101T000000_B00.kml)
2. fetching of assets
    + available from both [AWS S3](http://sentinel-pds.s3-website.eu-central-1.amazonaws.com/) storage and [Google](https://cloud.google.com/storage/docs/public-datasets/sentinel-2)
    + also, of course, available from ESA via their [SciHub](https://scihub.copernicus.eu/userguide/5APIsAndBatchScripting)
3. (tbd) processing of assets

**Comment by ircwaves** [comment]
Found my note about how to search for the asset names (and Download links) via ESA sentinel hub:

https://scihub.copernicus.eu/dhus/search?q=platformname:Sentinel-2&ingestiondate:[NOW-10DAY%20TO%20NOW-9DAY]

**Comment by matthewhanson** [comment]
Note that you can use the DevSeed API for Sentinel which may be easier, and provides the links to AWS.

https://github.com/sat-utils/sat-api

You can run your own instance of the API or use the one at
https://api.developmentseed.org/satellites

**Comment by ircwaves** [comment]
I'll check that out.  Looks like both AWS and GoogleCloudPlatform make the file contents available.  Do you know if there is there a param to download the whole SAFE as a zip?

The ESA API doesn't seem too bad:

https://scihub.copernicus.eu/dhus/search?q=platformname:Sentinel-2%20AND%20*_T42NYJ_*%20AND%20endposition:[2017-01-20T00:00:00.000Z%20TO%202017-01-21T00:00:00.000Z]

**Comment by matthewhanson** [comment]
you can use boto3 or the awscli to sync the entire folder, but not as a zip.  It wouldn't be any smaller anyway, they are in JP2K format already heavily compressed.

**Comment by ircwaves** [comment]
That makes sense.  Still dig that I see one of those funny looking dolphin in my mind ever time i see the word boto.

**Comment by ircwaves** [comment]
@ags-tolson here is the page to get a scihub account https://scihub.copernicus.eu/dhus/#/self-registration

**Comment by ags-tolson** [comment]
@ircwaves thanks!

**Comment by ircwaves** [comment]
https://github.com/Applied-GeoSolutions/gips/compare/dev...Applied-GeoSolutions:200-sentinel2-driver#diff-3c3a9e46a9ee84a31b87335c337e8590R129

This command needs a `--no-verbose` unless the `gippy.Options.Verbose()` is turned up to 3 or 4...because in a batched mode, the updating of the console maps into generating gobs of lines of text.

**Comment by ags-tolson** [comment]
`--no-verbose` is already on, so I think it's `--show-progress` that's the issue in this case.  It's supposed to produce clean output for both terminal and redirecting to a file, but I tried it just now and it's terribly verbose for a large file.

However, `--progress=dot:giga` works cleanly both in terminals and when redirecting to a file, and isn't too verbose even for files of this size.  Is leaving that on for `-v1` a bad idea, given that it's not too noisy?  The progress output per asset amounts to circa 12 lines.

**Comment by ircwaves** [comment]
`dot:giga` works...still gonna spew 400 dots, but not that bad for `-v1`...should be no progress meter at `-v0`.

**Comment by ags-tolson** [comment]
okay, done, pushed -> d7d6cc3

**Comment by ircwaves** [comment]
sub-task issue #208

**Comment by ags-tolson** [comment]
update for people watching who aren't ircwaves:  first product, NDVI-toa, is functional in the sentinel2-branch:

https://github.com/Applied-GeoSolutions/gips/tree/200-sentinel2-driver

**Comment by ags-tolson** [comment]
Current status since the big initial Sentinel-2 PR just merged (#220):  This issue's tasks are done, except for "(tbd) processing of assets".  Current products are:
* TOA Indices
* TOA ref product

Next I'm working towards atmo-corrected products, here's the path to that:
* First de-compute the sentinel-2 asset data from a TOA reflectance product into a TOA radiance product.
* Then feed radiance values into an atmo correction process similar to our other drivers (probably mostly like landsat).
* Provide code paths for letting the user choose TOA or atmo-corrected versions of each product.

**Comment by ircwaves** [comment]
Looking good.  Last steps:

1. finish `rad` product
2. devise plan for pre-2012-12-06 assets.

**Comment by ags-tolson** [comment]
Current state of dev:  `200-atmo-cor` implements rad-toa, surface rad, and surface ref products.   `200-support-old-assets` is based on that branch, and adds partial support for old-style assets.  I can't remember the state of system tests in either of those branches.

**Comment by ags-tolson** [comment]
You're welcome!  But wait, there's more!  Here's what's left regarding old-style assets:
* fetching is supported, and archiving is drafted, but untested
* product generation may be partly supported; ref-toa has been generated successfully using a hand-placed asset
* possibly adding `@skip` to the system tests so they won't mess with CI while we're waiting on verification from the science team (a driver comes pre-feature-toggled in gips, so it's harmless to merge untested if you want).

**Comment by ags-tolson** [comment]
Old-style assets support de-duplication now (as two different fetches may yield the same asset download); next is to get into process() and produce products.

CF https://github.com/Applied-GeoSolutions/gips/projects/5

**Comment by ags-tolson** [comment]
@ircwaves you had asked me for an estimate for finishing old asset support:  I looked at the code this AM and believe it's 3-5 days.

**Comment by ags-tolson** [comment]
Current status:  old-style assets are supported in this branch: https://github.com/Applied-GeoSolutions/gips/tree/200-support-old-assets

ref-toa, rad-toa, and toa indices look like they work based on inspecting output tifs.  atmo-corrected products (rad, ref, and non-toa indices) are implemented but untested at the moment due to some unique dependencies.

**Comment by ircwaves** [comment]
> unique dependencies.

That's the AOD business, right?

**Comment by ags-tolson** [comment]
@ircwaves yes indeed; for now I have some automated testing to clean up, then if time remains for today I have another issue in my queue.

**Comment by ags-tolson** [comment]
Merged into dev, everything seems to work, closing.
